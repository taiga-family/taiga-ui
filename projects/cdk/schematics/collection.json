{
    "$schema": "../../../node_modules/@angular-devkit/schematics/collection-schema.json",
    "schematics": {
        "ng-add": {
            "description": "Add Taiga UI to the project.",
            "factory": "./ng-add/index#ngAdd",
            "schema": "./ng-add/schema.json"
        },
        "ng-add-setup-project": {
            "private": true,
            "description": "Sets up the project after dependencies have been installed",
            "factory": "./ng-add/setup-project",
            "schema": "./ng-add/schema.json"
        },
        "migrateIconsV4": {
            "private": true,
            "description": "Migrate Taiga UI icons to a new format",
            "factory": "./ng-update/v4/migrate-icons/index#migrateIcons",
            "schema": "./ng-update/v4/schema.json"
        },
        "migrateCssVarsV4": {
            "private": true,
            "description": "Migrate Taiga UI css variables",
            "factory": "./ng-update/v4/migrate-css-vars/index#migrateCssVars",
            "schema": "./ng-update/v4/schema.json"
        },
        "updateToV4": {
            "private": true,
            "description": "Manually run migration to update Taiga to v4.x.x",
            "factory": "./ng-update/v4/index#updateToV4",
            "schema": "./ng-update/v4/schema.json"
        },
        "migrateTuiLet": {
            "description": "Migrate TuiLet directive to @let control flow",
            "factory": "./migrate-tui-let/index",
            "schema": "./ng-add/schema.json"
        }
    }
}
