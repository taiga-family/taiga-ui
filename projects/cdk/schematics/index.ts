export * from './constants/file-globs';
export * from './constants/taiga-styles';
export * from './ng-add/constants/modules';
export * from './ng-add/constants/packages';
export * from './ng-add/constants/versions';
export type * from './ng-add/schema';
export * from './utils/add-import-to-closest-module';
export * from './utils/add-unique-import';
export * from './utils/angular-json-manipulations';
export * from './utils/create-angular-json';
export * from './utils/get-component-from-identifier';
export * from './utils/get-execution-time';
export * from './utils/get-named-import-references';
export * from './utils/get-project-target-options';
export * from './utils/get-projects';
export * from './utils/get-standalone-bootstrap-function';
export * from './utils/import-manipulations';
export * from './utils/insert-todo';
export * from './utils/is-ci';
export * from './utils/progress';
export * from './utils/project-root';
export * from './utils/push-to-array-property';
