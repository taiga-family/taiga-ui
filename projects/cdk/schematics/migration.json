{
    "$schema": "../../../node_modules/@angular-devkit/schematics/collection-schema.json",
    "schematics": {
        "migrateCssVarsV4": {
            "description": "Migrate Taiga UI css variables",
            "version": "4.0.0",
            "factory": "./ng-update/v4/migrate-css-vars/index#migrateCssVars"
        },
        "migrateIconsV4": {
            "description": "Migrate Taiga UI icons to a new format",
            "version": "4.0.0",
            "factory": "./ng-update/v4/migrate-icons/index#migrateIcons"
        },
        "migrateAddonDocV4": {
            "description": "Migrate Taiga UI addon doc",
            "version": "4.0.0",
            "factory": "./ng-update/v4/migrate-addon-doc/index#migrateAddonDoc"
        },
        "updateToV4": {
            "description": "Updates Taiga UI packages to v4",
            "version": "4.0.0",
            "factory": "./ng-update/v4/index#updateToV4"
        },
        "updateToV5": {
            "description": "Updates Taiga UI packages to v5",
            "version": "5.0.0",
            "factory": "./ng-update/v5/index#updateToV5"
        }
    }
}
