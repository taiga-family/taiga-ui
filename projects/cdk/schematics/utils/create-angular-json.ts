import {createSourceFile} from 'ng-morph';

export function createAngularJson({
    stylesExist = false,
    root = '',
}: {stylesExist?: boolean; root?: string} = {}): void {
    createSourceFile(
        'angular.json',
        `
{
  "version": 1,
  "projects": {
    "demo": {
        "root": "${root}",
        "architect": {
          "build": {
            "options": {
              "main": "test/main.ts",
            ${
                stylesExist
                    ? `"styles": [
                  "node_modules/@taiga-ui/core/styles/taiga-ui-theme.less",
                  "some.style"
                ]
                `
                    : ''
            }}
          }
        }
    }
  }
}`,
        {overwrite: true},
    );
}
