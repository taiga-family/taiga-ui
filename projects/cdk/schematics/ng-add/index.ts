import {type Rule, type SchematicContext, type Tree} from '@angular-devkit/schematics';
import {NodePackageInstallTask, RunSchematicTask} from '@angular-devkit/schematics/tasks';
import {
    addPackageJsonDependency,
    getPackageJsonDependency,
    NodeDependencyType,
    removePackageJsonDependency,
} from 'ng-morph';

import {MAIN_PACKAGES} from './constants/packages';
import {TAIGA_VERSION} from './constants/versions';
import {type TuiSchema} from './schema';

function addDependencies(tree: Tree, options: TuiSchema): void {
    const packages = [...MAIN_PACKAGES, ...options.addons];

    packages.forEach((pack) => {
        addPackageJsonDependency(tree, {
            name: `@taiga-ui/${pack}`,
            version: TAIGA_VERSION,
            type: NodeDependencyType.Default,
        });
    });

    addPackageJsonDependency(tree, {
        name: '@taiga-ui/event-plugins',
        version: '^4.0.2',
    });

    removeTaigaSchematicsPackage(tree);

    if (packages.includes('addon-table') || packages.includes('addon-mobile')) {
        addAngularCdkDep(tree);
    }
}

function addAngularCdkDep(tree: Tree): void {
    const angularCore = getPackageJsonDependency(tree, '@angular/core')?.version;

    if (!angularCore) {
        return;
    }

    const majorVersionArr = /\d+/.exec(angularCore);

    if (majorVersionArr) {
        addPackageJsonDependency(tree, {
            name: '@angular/cdk',
            version: `^${majorVersionArr[0]}.0.0`,
            type: NodeDependencyType.Default,
        });
    }
}

function removeTaigaSchematicsPackage(tree: Tree): void {
    try {
        if (getPackageJsonDependency(tree, 'taiga-ui')?.version) {
            removePackageJsonDependency(tree, 'taiga-ui');
        }
    } catch {
        // noop
    }
}

export function ngAdd(options: TuiSchema): Rule {
    return (tree: Tree, context: SchematicContext): void => {
        context.logger.info(`The main packages will be installed - ${MAIN_PACKAGES}`);

        addDependencies(tree, options);

        context.addTask(new NodePackageInstallTask(), [
            context.addTask(new RunSchematicTask('ng-add-setup-project', options)),
        ]);
    };
}
