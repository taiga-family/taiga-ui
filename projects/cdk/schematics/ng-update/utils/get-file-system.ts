import {type Tree} from '@angular-devkit/schematics';
import {createProject, type DevkitFileSystem, setActiveProject} from 'ng-morph';

import {ALL_FILES} from '../../constants';
import {projectRoot} from '../../utils/project-root';

export function getFileSystem(tree: Tree): DevkitFileSystem {
    const project = createProject(tree, projectRoot(), ALL_FILES);

    setActiveProject(project);

    return project.getFileSystem().fs;
}
