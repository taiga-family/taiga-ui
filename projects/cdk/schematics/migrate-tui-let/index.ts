import {type Rule, type Tree} from '@angular-devkit/schematics';
import {getWorkspace} from '@schematics/angular/utility/workspace';
import {createProject, errorLog, saveActiveProject, setActiveProject} from 'ng-morph';

import {ALL_FILES} from '../constants';
import {type TuiSchema} from '../ng-add/schema';
import {getProjects} from '../utils/get-projects';
import {removeModule} from '../utils/remove-module';
import {findElementsWithAttribute} from '../utils/templates/elements';
import {getComponentTemplates} from '../utils/templates/get-component-templates';
import {
    getPathFromTemplateResource,
    getTemplateFromTemplateResource,
    getTemplateOffset,
} from '../utils/templates/template-resource';

export default function tuiLetMigrationGenerator(options: TuiSchema): Rule {
    return async (tree: Tree): Promise<void> => {
        const workspace = await getWorkspace(tree);
        const [project] = getProjects(options, workspace);

        const root = project?.root ?? project?.sourceRoot;

        if (root === undefined) {
            !options['skip-logs'] &&
                errorLog('[ERROR]: Target project not found in current workspace');

            return;
        }

        const prj = createProject(tree, root, ALL_FILES);
        const fileSystem = prj.getFileSystem().fs;

        setActiveProject(prj);

        removeModule('TuiLet', '@taiga-ui/cdk');

        const resources = getComponentTemplates(ALL_FILES);

        for (const resource of resources) {
            const path = fileSystem.resolve(getPathFromTemplateResource(resource));
            const recorder = fileSystem.edit(path);
            const template = getTemplateFromTemplateResource(resource, fileSystem);
            const templateOffset = getTemplateOffset(resource);
            const attributeName = '*tuiLet';

            const elements = findElementsWithAttribute(template, attributeName);

            for (const {sourceCodeLocation, attrs, tagName, childNodes} of elements) {
                const {name, value} = attrs.find(
                    (attr) => attr.name === attributeName.toLowerCase(),
                )!;

                const indent =
                    sourceCodeLocation!.startOffset -
                    (template.lastIndexOf('\n', sourceCodeLocation!.startOffset) + 1);
                const indentStr = ' '.repeat(indent);

                const [expr, key] = value.split(' as ').map((c) => c.trim());

                if (new RegExp(`@let\\s+${key}\\s+=`).test(template)) {
                    !options['skip-logs'] &&
                        errorLog(`The @let with key ${key} is already defined`);

                    continue;
                }

                recorder.insertLeft(
                    templateOffset + sourceCodeLocation!.startOffset,
                    `@let ${key} = ${expr};\n${indentStr}`,
                );

                if (tagName === 'ng-container' && attrs.length === 1) {
                    if (childNodes.length) {
                        const firstNode = childNodes.at(0)!;
                        const lastNode = childNodes.at(0)!;

                        recorder.remove(
                            templateOffset + sourceCodeLocation!.startOffset,
                            firstNode.sourceCodeLocation!.startOffset -
                                sourceCodeLocation!.startOffset,
                        );
                        recorder.remove(
                            templateOffset + lastNode.sourceCodeLocation!.endOffset,
                            sourceCodeLocation!.endOffset -
                                lastNode.sourceCodeLocation!.endOffset,
                        );
                    } else {
                        recorder.remove(
                            templateOffset + sourceCodeLocation!.startOffset,
                            sourceCodeLocation!.endOffset -
                                sourceCodeLocation!.startOffset,
                        );
                    }
                } else {
                    const {startOffset, endOffset} = sourceCodeLocation!.attrs![name]!;

                    recorder.remove(
                        templateOffset + startOffset - 1,
                        endOffset - startOffset + 1,
                    );
                }
            }
        }

        fileSystem.commitEdits();
        saveActiveProject();
    };
}
