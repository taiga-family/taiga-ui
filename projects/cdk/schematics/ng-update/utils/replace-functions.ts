import {Node} from 'ng-morph';

import {getNamedImportReferences} from '../../utils/get-named-import-references';
import {type ReplacementType} from '../interfaces/replacement-type';

export function replaceFunctions(functions: readonly ReplacementType[]): void {
    functions.forEach(({from, to, moduleSpecifier}) => {
        getNamedImportReferences(from, moduleSpecifier).forEach((ref) => {
            if (ref.wasForgotten()) {
                return;
            }

            const parent = ref.getParent();

            if (Node.isImportSpecifier(parent) || Node.isCallExpression(parent)) {
                parent?.replaceWithText(
                    parent
                        ?.getText({includeJsDocComments: false})
                        .trim()
                        .replace(from, to ?? from),
                );
            }
        });
    });
}
