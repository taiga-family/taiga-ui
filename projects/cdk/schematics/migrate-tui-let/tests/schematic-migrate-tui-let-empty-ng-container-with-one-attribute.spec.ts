import {join} from 'node:path';

import {HostTree} from '@angular-devkit/schematics';
import {SchematicTestRunner, UnitTestTree} from '@angular-devkit/schematics/testing';
import type {TuiSchema} from '@taiga-ui/cdk/schematics/ng-add/schema';
import {
    createProject,
    createSourceFile,
    resetActiveProject,
    saveActiveProject,
    setActiveProject,
} from 'ng-morph';

import {createAngularJson} from '../../utils/create-angular-json';

const collectionPath = join(__dirname, '../../collection.json');

const COMPONENT = `
import { TuiLet } from "@taiga-ui/cdk";

@Component({
    standalone: true,
    templateUrl: './test.template.html',
    imports: [TuiLet],
})
export class Test {
    readonly value = 'foo';
}`;

const TEMPLATE = `
<ng-container *tuiLet="value as val">
</ng-container>

<div>
    <ng-container *tuiLet="value as val2">
    </ng-container>
</div>
`;

describe('migrateTuiLet', () => {
    let host: UnitTestTree;
    let runner: SchematicTestRunner;

    beforeEach(() => {
        host = new UnitTestTree(new HostTree());
        runner = new SchematicTestRunner('schematics', collectionPath);

        setActiveProject(createProject(host));

        createMainFiles();

        saveActiveProject();
    });

    it('should migrate template', async () => {
        const tree = await runner.runSchematic(
            'migrateTuiLet',
            {'skip-logs': process.env['TUI_CI'] === 'true'} as Partial<TuiSchema>,
            host,
        );

        expect(tree.readContent('test/app/test.template.html')).toMatchSnapshot();
    });

    it('should migrate ts files', async () => {
        const tree = await runner.runSchematic(
            'migrateTuiLet',
            {'skip-logs': process.env['TUI_CI'] === 'true'} as Partial<TuiSchema>,
            host,
        );

        expect(tree.readContent('test/app/test.component.ts')).toMatchSnapshot();
    });

    afterEach(() => {
        resetActiveProject();
    });
});

function createMainFiles(): void {
    createSourceFile('test/app/test.component.ts', COMPONENT);

    createSourceFile('test/app/test.template.html', TEMPLATE);

    createAngularJson();
}
