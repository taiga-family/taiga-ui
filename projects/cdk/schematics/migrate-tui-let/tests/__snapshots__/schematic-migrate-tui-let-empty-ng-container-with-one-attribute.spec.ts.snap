// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrateTuiLet should migrate template 1`] = `
"
@let val = value;



<div>
    @let val2 = value;
    
    
</div>
"
`;

exports[`migrateTuiLet should migrate ts files 1`] = `
"
@Component({
    standalone: true,
    templateUrl: './test.template.html',
    imports: [],
})
export class Test {
    readonly value = 'foo';
}"
`;
